File: api/banana.rb
# ~/app/server.rb
# A simple WS server that evaluates math expression literally using ruby eval
# Run this server with ruby server.rb

require "rack"
require "rack/handler/puma"
require "lite_cable"
require "lite_cable/server"

class CalculatorConnection < LiteCable::Connection::Base
  identified_by :token

  def connect
    self.token = request.params["token"]

    $stdout.puts "Session #{token} is connected to the server"
  end

  def disconnect
    $stdout.puts "Session #{token} disconnected from the server"
  end
end

class CalculatorChannel < LiteCable::Channel::Base
  identifier :calculator

  def subscribed
    stream_from "calculator"
  end

  def evaluate(payload)
    $stdout.puts "Data received from the client"

    LiteCable.broadcast "calculator", result: eval(payload['expression'])
  end
end

app = Rack::Builder.new
app.map "/cable" do
  use LiteCable::Server::Middleware, connection_class: CalculatorConnection

  run(proc { |_| [200, {"Content-Type" => "text/plain"}, ["OK"]] })
end

Rack::Handler::Puma.run app

File: api/banana.js
// ~/app/client.js
// A client that talks to the WS server.
// To run, Open chrome console and paste this code

let socket = null;

(function init() {
  socket = new WebSocket("ws://0.0.0.0:9292/cable?token=abc123");

  socket.onopen = function (e) {
    console.info("[connected] : Connection established to the server");
  };

  socket.onmessage = function (e) {
    console.info("[message] : Message from the server", e.data);
  };

  socket.onerror = function (e) {
    console.error("[error] : Error establishing connection", e.message);
  };

})();

// subscribe to the calculator channel
socket.send(
  JSON.stringify({
    command: "subscribe",
    identifier: JSON.stringify({
      channel: "calculator",
    }),
  })
);

// call the evaluate method on the Calculator channel
socket.send(
  JSON.stringify({
    command: "message",
    identifier: JSON.stringify({
      channel: "calculator",
    }),
    data: JSON.stringify({
      expression: "150 * 4 - 3 / (2 + 2)",
      action: "evaluate",
    }),
  })
);
